src/csrs/service/csrsService.ts

```diff
@@ -196,22 +196,37 @@ export class CsrsService {
 		return await this.restService.getWithConfig(reportUrl, this.getAuthorizationHeader(user))
 	}
 
-	async getOrganisationalUnitsForUser(user: any) {
+	async getOrganisationalUnitsForUser(user: any): Promise<OrganisationalUnit[]> {
 		let organisationList = await this.listOrganisationalUnitsForTypehead()
-		let organisationsForTypeahead = organisationList.typeahead
+		let allOrganisations = new Map<number, OrganisationalUnit> (organisationList.typeahead.map((o): [number, OrganisationalUnit] => [o.id, o]))
 
 		if(!this.userCanAccessAllOrganisations(user)){
-			organisationsForTypeahead = organisationsForTypeahead.filter((org) => org.domains.map(domain => domain.domain).includes(user.getDomain()))
-			if(user.isTierOneReporter()){
-				const userTierOneOrganisation: OrganisationalUnit = user.organisationalUnit.getTopTierOrganisation(organisationsForTypeahead)
-				organisationsForTypeahead = organisationsForTypeahead.filter(organisation => !organisation.isTierOneOrganisation() || organisation.id === userTierOneOrganisation.id)
-			}
-			else{
-				organisationsForTypeahead = organisationsForTypeahead.filter(organisation => !organisation.isTierOneOrganisation())
+			const filteredOrgs = []
+			const filteredOrgIds: number[] = []
+
+			for (const organisationalUnit of allOrganisations.values()) {
```

Comment:
Is it worth putting a debug statement inside the for loop to detect that it is not going into an infinite loop?

Reply:
I think this is fine, `allOrganisations.values()` is a finite number and not a while loop